File: crop-me/src/main/java/com/mikelau/croperino/BackgroundJob.java
package com.mikelau.croperino;

import android.app.ProgressDialog;
import android.os.Handler;

/**
 * Created by Mike on 9/14/2016.
 */
public class BackgroundJob extends MonitoredActivity.LifeCycleAdapter implements Runnable {

    private final MonitoredActivity mActivity;
    private final ProgressDialog mDialog;
    private final Runnable          mJob;
    private final Handler mHandler;
    private final Runnable mCleanupRunner = new Runnable() {
        public void run() {

            mActivity.removeLifeCycleListener(BackgroundJob.this);
            if (mDialog.getWindow() != null) mDialog.dismiss();
        }
    };

    public BackgroundJob(MonitoredActivity activity, Runnable job,
                         ProgressDialog dialog, Handler handler) {

        mActivity = activity;
        mDialog = dialog;
        mJob = job;
        mActivity.addLifeCycleListener(this);
        mHandler = handler;
    }

    public void run() {
        try {
            mJob.run();
        } finally {
            mHandler.post(mCleanupRunner);
        }
    }


    @Override
    public void onActivityDestroyed(MonitoredActivity activity) {
        mCleanupRunner.run();
        mHandler.removeCallbacks(mCleanupRunner);
    }

    @Override
    public void onActivityStopped(MonitoredActivity activity) {

        mDialog.hide();
    }

    @Override
    public void onActivityStarted(MonitoredActivity activity) {

        mDialog.show();
    }
}

File: croperino/src/main/java/com/mikelau/croperino/BackgroundJob.kt
package com.mikelau.croperino

import android.app.ProgressDialog
import android.os.Handler
import com.mikelau.croperino.MonitoredActivity.LifeCycleAdapter

class BackgroundJob(
    private val mActivity: MonitoredActivity,
    private val mJob: Runnable,
    private val mDialog: ProgressDialog,
    handler: Handler
) : LifeCycleAdapter(), Runnable {
    private val mHandler: Handler
    private val mCleanupRunner = Runnable {
        mActivity.removeLifeCycleListener(this@BackgroundJob)
        if (mDialog.window != null) mDialog.dismiss()
    }

    init {
        mActivity.addLifeCycleListener(this)
        mHandler = handler
    }

    override fun run() {
        try {
            mJob.run()
        } finally {
            mHandler.post(mCleanupRunner)
        }
    }

    override fun onActivityDestroyed(activity: MonitoredActivity?) {
        mCleanupRunner.run()
        mHandler.removeCallbacks(mCleanupRunner)
    }

    override fun onActivityStopped(activity: MonitoredActivity?) {
        mDialog.hide()
    }

    override fun onActivityStarted(activity: MonitoredActivity?) {
        mDialog.show()
    }
}
